akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -172,29 +108,16 @@ private[akka] class OutputStreamAdapter(
   private[this] def sendData(data: ByteString): Unit =
     send(() ⇒ {
       try {
-        dataQueue.put(data)
+        // FIXME take into account writeTimeout here.
+        unfulfilledDemand.acquire()
+        sendToStage(Send(data))
       } catch { case NonFatal(ex) ⇒ throw new IOException(ex) }
       if (downstreamStatus.get() == Canceled) {
         isPublisherAlive = false
         throw publisherClosedException
       }
```

Comment:
Existing logic, but, is it really worth saving four lines of code at the cost of allocating a lambda on each `sendData` call? (the send param) I think that it could be expressed imperatively and be reused instead, something like 
```
throwIfCannotSend()
doStuffBecauseCanSend() ...
```

Reply:
This can even evaporate entirely with invokeWithFeedback